src/libcore/pin.rs

```diff
@@ -149,22 +149,23 @@
 //! of your type could have been pinned, you must treat Drop as implicitly taking
 //! `Pin<&mut Self>`.
 //!
+//! In particular, if your type is `#[repr(packed)]`, the compiler will automatically
+//! move fields around to be able to drop them. As a consequence, you cannot use
+//! pinning with a `#[repr(packed)]` type.
+//!
 //! # Projections and Structural Pinning
 //!
 //! One interesting question arises when considering the interaction of pinning and
-//! the fields of a struct. When can a struct have a "projection operation", i.e.,
+//! the fields of a struct. When can a struct have a "pinning projection", i.e.,
 //! an operation with type `fn(Pin<&[mut] Struct>) -> Pin<&[mut] Field>`?
 //! In a similar vein, when can a container type (such as `Vec`, `Box`, or `RefCell`)
 //! have an operation with type `fn(Pin<&[mut] Container<T>>) -> Pin<&[mut] T>`?
 //!
 //! This question is closely related to the question of whether pinning is "structural":
-//! when you have pinned a wrapper type, have you pinned its contents? Adding a
+//! when you have pinned a wrapper type, have you pinned its contents? Deciding this
+//! is entirely up to the author of any given type. However, adding a
```

Comment:
I think an example of when you might decide that its contents aren't pinned even though the wrapper is would be really helpful here. I'm having a hard time wrapping my head around it.

Reply:
Further down, we say that the standard library does this for all sorts of types.

But also, clearly, if you cannot satisfy one of the requirements given in the list, that might be a good enough reason -- right?

It is in many cases really arbitrary. For many types such as `Vec`, you could literally have two versions that differ only in whether they pin their content or not, and then one would have projections and the other would have `impl<T> Unpin for Vec<T>`. Really, it's up to you as the author whether you consider projections or free `Unpin` instances more viable.

Reply:
I think this is resolved. I think my thought process here was along the lines of: why would it ever be the case that if the structure can't move, its fields can't move? And the answer is that it might internally move out of its fields (like `Option::take`), which wouldn't be okay. I still think mentioning something like an `Option` field as an example here would help guide the reader's thinking, but omitting it is probably also okay.

Reply:
> why would it ever be the case that if the structure can't move, its fields can't move?

I assume you mean "can move" at the end of the sentence?

Reply:
I have tried to reword this text to make it clearer. Does that help?

Reply:
Sorry, yes, that's what I meant. I feel like the text still assumes that the reader understands what you mean by "structural pinning", whereas without an example I think that's actually quite hard to tease out. You give a definition, but without an example of what would and would not meet the requirements of structural pinning, I think the following text will be hard to follow. To me at least, giving the very concrete example of:

> For example, if a wrapper type contains an `Option<T>` whose value is taken (e.g., with `Option::take`), then even though the wrapper type is pinned, the `T` inside the `Option` field is _not_ pinned (its location may change), and it can therefore not safely be exposed.

Would help immensely by giving a concrete, guiding mental example.

Reply:
> without an example of what would and would not meet the requirements of structural pinning, I think the following text will be hard to follow. To me at least, giving the very concrete example of:

The following 4-point list is exactly meant to explain that, and it has some examples. I added the `Option` example there, and also changed the text a bit.

I feel like adding an example before the list makes the entire structure very messy and hard to follow.

Reply:
Mmm... I see what you mean. The updated text looks good :+1: